### PR TITLE
Move from {{closure}}#0 syntax to {closure#0} for (def) path components

### PR DESCRIPTION
**Pull request by marmeladema**
Part of #70334

I followed the approach described by @eddyb and introduced a `DefPathDataName` enum.
To preserve compatibility, in various places, I had to rely on formatting manually by calling `format!("{{{{{}}}}}", namespace)`.

My questions are:
* Do we want to convert for places to use the new naming scheme? Or shall I re-add `DefPathData::as_symbol` but renamed as `DefPathData::as_legacy_symbol` to avoid manually allocating the legacy symbols?
* Do we want to `impl Display for DisambiguatedDefPathData` to avoid manually calling `write!(s, "{{{}#{}}}", namespace, component.disambiguator)`?
* We might also want to improve naming for `DefPathDataName` and `DefPathData::get_name`

r? @eddyb

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/debuginfo/namespace.rs [hunk 1]
```diff
@@ -29,7 +30,12 @@ pub fn item_namespace(cx: &CodegenCx<'ll, '_>, def_id: DefId) -> &'ll DIScope {
 
     let namespace_name = match def_key.disambiguated_data.data {
         DefPathData::CrateRoot => cx.tcx.crate_name(def_id.krate),
-        data => data.as_symbol(),
+        data => match data.get_name() {
+            DefPathDataName::Named(name) => name,
+            DefPathDataName::Anon { namespace } => {
+                Symbol::intern(&format!("{{{{{}}}}}", namespace))
```
Comment (eddyb) [inline compiler/rustc_codegen_llvm/src/debuginfo/namespace.rs:36]: Can you add a `FIXME` comment on each of these? And mention that it produces the old form "e.g. `{{closure}}`".

File: compiler/rustc_hir/src/definitions.rs [hunk 2]
```diff
@@ -202,7 +202,12 @@ impl DefPath {
         let mut s = String::with_capacity(self.data.len() * 16);
 
         for component in &self.data {
-            write!(s, "::{}[{}]", component.data.as_symbol(), component.disambiguator).unwrap();
+            match component.data.get_name() {
+                DefPathDataName::Named(name) => write!(s, "::{}", name).unwrap(),
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:206]: This was always printing the disambiguator so you should also print it with `#` for the `Named` case (but you can limit it to the case where `disambiguator != 0` if you want).

File: compiler/rustc_hir/src/definitions.rs [hunk 3]
```diff
@@ -220,10 +225,11 @@ impl DefPath {
         write!(s, "::{}", crate_name_str).unwrap();
 
         for component in &self.data {
-            if component.disambiguator == 0 {
-                write!(s, "::{}", component.data.as_symbol()).unwrap();
-            } else {
-                write!(s, "{}[{}]", component.data.as_symbol(), component.disambiguator).unwrap();
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:226]: This not printing `::` for the `disambiguator != 0` case looks like a bug, I don't know why that would be.

File: compiler/rustc_hir/src/definitions.rs [hunk 4]
```diff
@@ -220,10 +225,11 @@ impl DefPath {
...
+            match component.data.get_name() {
+                DefPathDataName::Named(name) => write!(s, "::{}", name).unwrap(),
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:229]: This also needs the disambiguator when not `0`.

File: compiler/rustc_hir/src/definitions.rs [hunk 5]
```diff
@@ -240,10 +246,11 @@ impl DefPath {
         for component in &self.data {
             s.extend(opt_delimiter);
             opt_delimiter = Some('-');
-            if component.disambiguator == 0 {
-                write!(s, "{}", component.data.as_symbol()).unwrap();
-            } else {
-                write!(s, "{}[{}]", component.data.as_symbol(), component.disambiguator).unwrap();
+            match component.data.get_name() {
+                DefPathDataName::Named(name) => write!(s, "{}", name).unwrap(),
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:250]: This also needs the disambiguator when not `0`.

File: compiler/rustc_hir/src/definitions.rs [hunk 6]
```diff
@@ -436,22 +448,27 @@ impl DefPathData {
         }
     }
 
-    pub fn as_symbol(&self) -> Symbol {
+    pub fn get_name(&self) -> DefPathDataName {
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:451]: We tend to avoid `get_` prefixes, this can just be `name`.

File: compiler/rustc_hir/src/definitions.rs [hunk 7]
```diff
@@ -436,22 +448,27 @@ impl DefPathData {
...
-            CrateRoot => sym::double_braced_crate,
-            Impl => sym::double_braced_impl,
-            Misc => sym::double_braced_misc,
-            ClosureExpr => sym::double_braced_closure,
-            Ctor => sym::double_braced_constructor,
-            AnonConst => sym::double_braced_constant,
-            ImplTrait => sym::double_braced_opaque,
+            CrateRoot => DefPathDataName::Anon { namespace: kw::Crate },
+            Impl => DefPathDataName::Anon { namespace: kw::Impl },
+            Misc => DefPathDataName::Anon { namespace: sym::misc },
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:460]: We might want to get rid of `Misc`, but that's not really relevant to this PR.

File: compiler/rustc_hir/src/definitions.rs [hunk 8]
```diff
@@ -436,22 +448,27 @@ impl DefPathData {
...
+            ClosureExpr => DefPathDataName::Anon { namespace: sym::closure },
+            Ctor => DefPathDataName::Anon { namespace: sym::constructor },
+            AnonConst => DefPathDataName::Anon { namespace: sym::constant },
+            ImplTrait => DefPathDataName::Anon { namespace: sym::opaque },
         }
     }
 
     pub fn to_string(&self) -> String {
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:468]: Hmm this probably wants to be a `fmt::Display` impl, although, I wonder where it's used (since we'd want to avoid calling `.to_string()` in most cases *anyway*).

File: compiler/rustc_lint/src/context.rs [hunk 9]
```diff
@@ -835,7 +835,12 @@ impl<'tcx> LateContext<'tcx> {
                     return Ok(path);
                 }
 
-                path.push(disambiguated_data.data.as_symbol());
+                path.push(match disambiguated_data.data.get_name() {
+                    DefPathDataName::Named(name) => name,
+                    DefPathDataName::Anon { namespace } => {
+                        Symbol::intern(&format!("{{{{{}}}}}", namespace))
+                    }
```
Comment (eddyb) [inline compiler/rustc_lint/src/context.rs:842]: Hmm, if you're doing this just to avoid the perf hit in the other case, you could still do `name => Symbol::intern(&name.to_string())`.

But also, ideally lints using this would just get an empty `Vec` back when the path contains anything other than simple names (see comment on `fn get_def_path`, about 100 lines above, on how it's meant to be used).

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 10]
```diff
@@ -1445,25 +1445,26 @@ impl<F: fmt::Write> Printer<'tcx> for FmtPrinter<'_, 'tcx, F> {
...
             }
-            write!(self, "{}", name)?;
 
-            // FIXME(eddyb) this will print e.g. `{{closure}}#3`, but it
-            // might be nicer to use something else, e.g. `{closure#3}`.
-            let dis = disambiguated_data.disambiguator;
-            let print_dis = disambiguated_data.data.get_opt_name().is_none()
-                || dis != 0 && self.tcx.sess.verbose();
-            if print_dis {
-                write!(self, "#{}", dis)?;
```
Comment (eddyb) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1466]: You need to keep existing "should the disambiguator be printed" logic for the `Named` case.

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 11]
```diff
@@ -1445,25 +1445,26 @@ impl<F: fmt::Write> Printer<'tcx> for FmtPrinter<'_, 'tcx, F> {
...
         // FIXME(eddyb) `name` should never be empty, but it
         // currently is for `extern { ... }` "foreign modules".
-        let name = disambiguated_data.data.as_symbol();
         if name != kw::Invalid {
             if !self.empty_path {
                 write!(self, "::")?;
             }
             if Ident::with_dummy_span(name).is_raw_guess() {
                 write!(self, "r#")?;
             }
```
Comment (eddyb) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1461]: This can be moved into the `Named` case.

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 12]
```diff
@@ -1445,25 +1445,26 @@ impl<F: fmt::Write> Printer<'tcx> for FmtPrinter<'_, 'tcx, F> {
             return Ok(self);
         }
 
+        let name = match disambiguated_data.data.get_name() {
+            DefPathDataName::Named(name) => name,
+            DefPathDataName::Anon { namespace } => namespace,
+        };
```
Comment (eddyb) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1451]: This is somewhat unnecessary, you can just rewrite the use below to check for `DefPathData::Named(kw::Invalid)`.

File: compiler/rustc_mir/src/interpret/intrinsics/type_name.rs [hunk 13]
```diff
@@ -134,7 +134,12 @@ impl<'tcx> Printer<'tcx> for AbsolutePathPrinter<'tcx> {
 
         self.path.push_str("::");
 
-        self.path.push_str(&disambiguated_data.data.as_symbol().as_str());
+        match disambiguated_data.data.get_name() {
+            DefPathDataName::Named(name) => self.path.write_str(&name.as_str()).unwrap(),
+            DefPathDataName::Anon { namespace } => {
+                write!(self.path, "{{{{{}}}}}", namespace).unwrap()
+            }
+        }
```
Comment (eddyb) [inline compiler/rustc_mir/src/interpret/intrinsics/type_name.rs:142]: This would benefit from `fmt::Display`, although, this seems to not want disambiguator.

Anyway, this one you can change from `{{closure}}` to `{closure}`, nothing should be relying on it (unlike legacy symbol names, which might be parsed) - presumably you want to do observable changes like that in a separate PR?

File: compiler/rustc_mir/src/monomorphize/partitioning/default.rs [hunk 14]
```diff
@@ -354,7 +355,12 @@ fn compute_codegen_unit_name(
     *cache.entry((cgu_def_id, volatile)).or_insert_with(|| {
         let def_path = tcx.def_path(cgu_def_id);
 
-        let components = def_path.data.iter().map(|part| part.data.as_symbol());
+        let components = def_path.data.iter().map(|part| match part.data.get_name() {
+            DefPathDataName::Named(name) => name,
+            DefPathDataName::Anon { namespace } => {
+                Symbol::intern(&format!("{{{{{}}}}}", namespace))
+            }
```
Comment (eddyb) [inline compiler/rustc_mir/src/monomorphize/partitioning/default.rs:362]: Interestingly, it should be impossible to hit this, because of the logic above, requiring only `DefKind::Mod` components.

File: compiler/rustc_middle/src/ty/query/profiling_support.rs [hunk 15]
```diff
@@ -66,17 +66,25 @@ impl<'p, 'c, 'tcx> QueryKeyStringBuilder<'p, 'c, 'tcx> {
...
-                    end_index = 3;
-                } else {
-                    write!(&mut dis_buffer[..], "[{}]", def_key.disambiguated_data.disambiguator)
+                name = match other.get_name() {
+                    DefPathDataName::Named(name) => {
+                        dis = "";
+                        end_index = 3;
+                        name
+                    }
+                    DefPathDataName::Anon { namespace } => {
```
Comment (eddyb) [inline compiler/rustc_middle/src/ty/query/profiling_support.rs:75]: The distinction was made on `disambiguator == 0` before, this doesn't look equivalent (it also doesn't change the output, but that's orthogonal).

File: compiler/rustc_middle/src/ty/query/profiling_support.rs [hunk 16]
```diff
@@ -66,17 +66,25 @@ impl<'p, 'c, 'tcx> QueryKeyStringBuilder<'p, 'c, 'tcx> {
...
+                        )
                         .unwrap();
-                    let end_of_dis = dis_buffer.iter().position(|&c| c == b']').unwrap();
-                    dis = std::str::from_utf8(&dis_buffer[..end_of_dis + 1]).unwrap();
-                    end_index = 4;
-                }
+                        let end_of_dis = dis_buffer.iter().position(|&c| c == b']').unwrap();
+                        dis = std::str::from_utf8(&dis_buffer[..end_of_dis + 1]).unwrap();
+                        end_index = 4;
+                        namespace
```
Comment (eddyb) [inline compiler/rustc_middle/src/ty/query/profiling_support.rs:85]: Hmm I think this would result in `closure[123]` instead of `{{closure}}[123]` (before) or `{closure#123}` (that we want).

cc @wesleywiser @Mark-Simulacrum Do we have no way of testing profiling with query keys? It would be nice to have a before/after in this PR. But I guess bringing in e.g. `crox` (or something more custom based on `analyzeme`) would not be cheap (to build in the first place, I mean)? (unless we can reuse all of the `rustc`/tools dependencies? since I think most are already used by something else)

File: compiler/rustc_hir/src/definitions.rs (hunk 3 above)
Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs:226]: This function (`to_string_friendly`) is not actually used, I'll remove it entirely instead of fixing unused broken code.

File: compiler/rustc_middle/src/ty/print/pretty.rs (hunk 10 above)
Comment (marmeladema) [reply compiler/rustc_middle/src/ty/print/pretty.rs:1466]: To simplify the code, I've removed the check for `self.tcx.sess.verbose()` and now always print the disambiguator for the `Named` code if its `!= 0`

File: compiler/rustc_lint/src/context.rs (hunk 9 above)
Comment (marmeladema) [reply compiler/rustc_lint/src/context.rs:842]: I am not really doing this to avoid the perf hit but rather because `path` is a `Vec<Symbol>` so I need to dynamically create a symbol for the old naming scheme.
As mentioned in the PR description, I wonder if I should re-add a `DefPathData::as_legacy_symbol(&self) -> Symbol` method which would return the double braced namespace symbols?

File: compiler/rustc_middle/src/ty/print/pretty.rs (hunk 10 above)
Comment (marmeladema) [reply compiler/rustc_middle/src/ty/print/pretty.rs:1466]: Well, I can't really do that after all, it changes significantly some error messages around path with the same name.

File: compiler/rustc_lint/src/context.rs (hunk 9 above)
Comment (eddyb) [reply compiler/rustc_lint/src/context.rs:842]: > so I need to dynamically create a symbol for the old naming scheme

I meant the perf hit of doing that, even in the case where you can just extract the `Symbol` from the `Named` case`.

That is, you *could* just always do `Symbol::intern(&name.to_string())`, without special-casing `Named`, and it should still work AFAICT.

File: compiler/rustc_hir/src/definitions.rs [hunk 17]
```diff
@@ -450,8 +474,10 @@ impl DefPathData {
             ImplTrait => sym::double_braced_opaque,
         }
     }
+}
 
-    pub fn to_string(&self) -> String {
-        self.as_symbol().to_string()
+impl fmt::Display for DefPathData {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.as_legacy_symbol().as_str())
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:481]: Wait, how is this useful for a `fmt::Display`? It completely erases the disambiguator.

File: compiler/rustc_hir/src/definitions.rs [hunk 18]
```diff
@@ -436,7 +443,24 @@ impl DefPathData {
...
+            Impl => DefPathDataName::Anon { namespace: kw::Impl },
+            Misc => DefPathDataName::Anon { namespace: sym::misc },
+            ClosureExpr => DefPathDataName::Anon { namespace: sym::closure },
+            Ctor => DefPathDataName::Anon { namespace: sym::constructor },
+            AnonConst => DefPathDataName::Anon { namespace: sym::constant },
+            ImplTrait => DefPathDataName::Anon { namespace: sym::opaque },
+        }
+    }
+
+    pub fn as_legacy_symbol(&self) -> Symbol {
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:463]: Not super happy about keeping this around, it would be nice to not have `Symbol`s for the double-braced forms - they're not *really* needed for anything.

Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs:463]: What is better, calling `Symbol::intern` in the 3 or 4 places that needs a legacy symbol or keeping the old symbols around? I don't have a preference either way.

File: compiler/rustc_hir/src/definitions.rs [hunk 19]
```diff
@@ -155,6 +155,32 @@ pub struct DisambiguatedDefPathData {
     pub disambiguator: u32,
 }
 
+impl DisambiguatedDefPathData {
+    pub fn format<W: Write>(&self, writer: &mut W, verbose: bool) -> fmt::Result {
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:159]: There's a more "user-friendly" way to do this, that would have this signature:
```rust
fn display(&self, verbose: bool) -> impl fmt::Display
```

If that seems a bit much, I'd call this `fmt_maybe_verbose` or something and keep the current signature (though you can have `write: &mut impl Write` instead of the explicit generic).

File: compiler/rustc_hir/src/definitions.rs [hunk 20]
```diff
@@ -155,6 +155,32 @@ pub struct DisambiguatedDefPathData {
...
+        match self.data.name() {
+            DefPathDataName::Named(name) => {
+                if Ident::with_dummy_span(name).is_raw_guess() {
+                    writer.write_str("r#")?;
+                }
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:164]: Hmm should we maybe handle this in other cases too? Seems a bit odd to have the `fmt::Display` impl for `DefPathData` exist while not handling that aspect.

File: compiler/rustc_hir/src/definitions.rs (hunk 18 above)
Comment (eddyb) [reply compiler/rustc_hir/src/definitions.rs:463]: It seems better to call `Symbol::intern` in a situation that shouldn't even be hit.

You could even e.g. write `if let DefPathDataName::Anon {..} = data.name() { unreachable!() }` and run tests, for each in turn.

File: compiler/rustc_hir/src/definitions.rs (hunk 20 above)
Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs:164]: I CAN add this check also in `impl Display for DefPathData` but I am concerned that it would introduce a (breacking?) change for the legacy names?

File: compiler/rustc_middle/src/ty/query/profiling_support.rs (hunk 16 above)
Comment (marmeladema) [reply compiler/rustc_middle/src/ty/query/profiling_support.rs]: Should be fixed now

File: compiler/rustc_span/src/symbol.rs [hunk 21]
```diff
@@ -679,6 +675,8 @@ symbols! {
         minnumf32,
         minnumf64,
         mips_target_feature,
+        misc,
+        mmx_target_feature,
```
Comment (eddyb) [inline compiler/rustc_span/src/symbol.rs:679]: Is this an accidental result of the rebase?

File: compiler/rustc_hir/src/definitions.rs [hunk 22]
```diff
@@ -155,6 +155,32 @@ pub struct DisambiguatedDefPathData {
...
 }
 
+impl DisambiguatedDefPathData {
+    pub fn fmt_maybe_verbose(&self, writer: &mut impl Write, verbose: bool) -> fmt::Result {
+        match self.data.name() {
+            DefPathDataName::Named(name) => {
+                if Ident::with_dummy_span(name).is_raw_guess() {
+                    writer.write_str("r#")?;
+                }
+                if self.disambiguator == 0 || !verbose {
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:165]: I think this would be clearer if the "then" and "else" cases were flipped (so the condition would be `verbose && self.disambiguator != 0`).

File: compiler/rustc_codegen_llvm/src/debuginfo/namespace.rs [hunk 23]
```diff
@@ -29,7 +30,7 @@ pub fn item_namespace(cx: &CodegenCx<'ll, '_>, def_id: DefId) -> &'ll DIScope {
 
     let namespace_name = match def_key.disambiguated_data.data {
         DefPathData::CrateRoot => cx.tcx.crate_name(def_id.krate),
-        data => data.as_symbol(),
+        data => Symbol::intern(&data.to_string()),
```
Comment (eddyb) [inline compiler/rustc_codegen_llvm/src/debuginfo/namespace.rs:33]: Since this is cached, I think it's better to use `.to_string()` on the `crate_name` (since crates are rarer) than `Symbol::intern` everything else.

Or if you want you can probably make this as cheap as possible like this:

```rust
    let crate_name_as_str;
    let name_to_string;
    let namespace_name = match def_key.disambiguated_data.data {
        DefPathData::CrateRoot => {
            crate_name_as_str = cx.tcx.crate_name(def_id.krate).as_str();
            &crate_name_as_str
        }
        data => {
            name_to_string = data.to_string();
            &name_to_string
        }
    };
```

File: compiler/rustc_hir/src/definitions.rs (hunk 20 above)
Comment (eddyb) [reply compiler/rustc_hir/src/definitions.rs:164]: Do any tests change behavior when you do that? `is_raw_guess` should only return `true` for keywords, so at least this should be rare.

File: compiler/rustc_hir/src/definitions.rs [hunk 24]
```diff
@@ -437,22 +444,29 @@ impl DefPathData {
...
     }
+}
 
-    pub fn to_string(&self) -> String {
-        self.as_symbol().to_string()
+impl fmt::Display for DefPathData {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.name() {
+            DefPathDataName::Named(name) => f.write_str(&name.as_str()),
+            DefPathDataName::Anon { namespace } => write!(f, "{{{{{}}}}}", namespace),
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:469]: Leave a FIXME pointing to the issue, perhaps? I guess in that case I would also remove the "Fixes #..." in the PR description.

File: compiler/rustc_hir/src/definitions.rs [hunk 25]
```diff
@@ -202,29 +228,7 @@ impl DefPath {
...
-
-        write!(s, "::{}", crate_name_str).unwrap();
-
-        for component in &self.data {
-            if component.disambiguator == 0 {
-                write!(s, "::{}", component.data.as_symbol()).unwrap();
-            } else {
-                write!(s, "{}[{}]", component.data.as_symbol(), component.disambiguator).unwrap();
-            }
+            write!(s, "::{}", component).unwrap();
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:231]: `to_string_no_crate` is used in the `fmt::Debug` implementation of `DefId`, so I think this should use the verbose mode.

 It's not used in much else either, so you can add `_verbose` to its name.

File: compiler/rustc_hir/src/definitions.rs [hunk 26]
```diff
@@ -240,12 +244,9 @@ impl DefPath {
         for component in &self.data {
             s.extend(opt_delimiter);
             opt_delimiter = Some('-');
-            if component.disambiguator == 0 {
-                write!(s, "{}", component.data.as_symbol()).unwrap();
-            } else {
-                write!(s, "{}[{}]", component.data.as_symbol(), component.disambiguator).unwrap();
-            }
+            write!(s, "{}", component).unwrap();
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:247]: This also wants the verbose mode, as we don't want filenames to accidentally overlap.

File: compiler/rustc_middle/src/ty/query/profiling_support.rs [hunk 27]
```diff
@@ -66,7 +67,7 @@ impl<'p, 'c, 'tcx> QueryKeyStringBuilder<'p, 'c, 'tcx> {
                 end_index = 3;
             }
             other => {
-                name = other.as_symbol();
+                name = Symbol::intern(&other.to_string());
```
Comment (eddyb) [inline compiler/rustc_middle/src/ty/query/profiling_support.rs:70]: Similar situation here to the debuginfo one: the crate is less common, so it could take the `.to_string()` cost better than this would the `Symbol::intern` cost. And as always you can use variables to work around needing to do either.

File: compiler/rustc_hir/src/definitions.rs (hunk 25 above)
Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs:231]: The implementation of `fmt::Display` for `DisambiguatedDefPathData` is already verbose but i'll rename the method to add `_verbose` suffix

File: compiler/rustc_hir/src/definitions.rs (hunk 26 above)
Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs:247]: The implementation of `fmt::Display` for `DisambiguatedDefPathData` is already verbose so I think this is fine?

File: compiler/rustc_hir/src/definitions.rs (hunk 20 above)
Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs:164]: Adding this to `DefPathData` is making a _lot of_ tests fail to compile with `undefined reference` errors. Most of them are for:
```
   = note: /bin/ld: /home/adema/Sources/rust/build/x86_64-unknown-linux-gnu/test/ui/threads-sendsync/task-comm-13/a.task_comm_13.7rcbfp3g-cgu.7.rcgu.o: in function `core::ops::function::FnOnce::call_once{{vtable-shim}}':
           task_comm_13.7rcbfp3g-cgu.7:(.text._ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h07af120783683994E+0xef): undefined reference to `std::panicking::try::cleanup'
           collect2: error: ld returned 1 exit status
```
or similar.

File: compiler/rustc_hir/src/definitions.rs (hunk 24 above)
Comment (eddyb) [reply compiler/rustc_hir/src/definitions.rs:471]: The common "syntax" is `// FIXME(#70334) ...` btw.

File: compiler/rustc_hir/src/definitions.rs (hunk 26 above)
Comment (eddyb) [reply compiler/rustc_hir/src/definitions.rs:247]: Oops, I missed that part, I expected the opposite.

File: compiler/rustc_hir/src/definitions.rs (hunk 20 above)
Comment (eddyb) [reply compiler/rustc_hir/src/definitions.rs:164]: Ah, weird, guess it's fine then.

Comment (eddyb) [reply compiler/rustc_hir/src/definitions.rs:164]: Though we could go the other way and move that check to `ty::print::pretty`, which might be the only place that cares.

File: compiler/rustc_hir/src/definitions.rs (hunk 24 above)
Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs:471]: Fixed ;)

File: compiler/rustc_hir/src/definitions.rs (hunk 20 above)
Comment (marmeladema) [reply compiler/rustc_hir/src/definitions.rs]: I moved it, should be all good now!

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 28]
```diff
@@ -1498,24 +1498,20 @@ impl<F: fmt::Write> Printer<'tcx> for FmtPrinter<'_, 'tcx, F> {
...
+        let name = disambiguated_data.data.name();
+        if name != DefPathDataName::Named(kw::Invalid) {
             if !self.empty_path {
                 write!(self, "::")?;
             }
-            if Ident::with_dummy_span(name).is_raw_guess() {
+            if disambiguated_data
+                .data
+                .get_opt_name()
+                .map_or(false, |name| Ident::with_dummy_span(name).is_raw_guess())
```
Comment (eddyb) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1509]: It's probably cleaner to write this the "boring" way i.e.
```rust
if let DefPathDataName::Named(name) = name {
    if Ident::with_dummy_span(name).is_raw_guess() {
        write!(self, "r#")?;
    }
}
```

File: compiler/rustc_hir/src/definitions.rs [hunk 29]
```diff
@@ -437,22 +441,31 @@ impl DefPathData {
...
     }
+}
 
-    pub fn to_string(&self) -> String {
-        self.as_symbol().to_string()
+impl fmt::Display for DefPathData {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.name() {
+            DefPathDataName::Named(name) => f.write_str(&name.as_str()),
+            // FIXME(#70334): necessary to be compatible with legacy symbol mangling.
```
Comment (eddyb) [inline compiler/rustc_hir/src/definitions.rs:466]: There's no "compatibility" to worry about, we just want to land it separately. But I guess this phrasing isn't necessary wrong I guess.

**Review by eddyb — APPROVED** [review]
r=me modulo the remaining comment

**Review by eddyb — APPROVED** [review]
